topics/reference_guide/custom_language_support/language_server_protocol.md

```diff
@@ -0,0 +1,175 @@
+
+The [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) (LSP) is an open-standard protocol developed by Microsoft. It enables communication between development tools and Language Servers.
+Language Servers can provide language-specific features such as code completion, documentation, and formatting, which is far easier than implementing language support from scratch.
+It also reduces the need for constant maintenance and tracking of changes in relevant languages and tools, making it easier to bring consistent language support to various development environments.
+
+However, the canonical custom language support provided by IntelliJ Platform still offers a wider range of integration with IDE features than handling and presenting data provided by a Language Server.
+Therefore, the LSP approach shouldn't be considered as a replacement for the existing language API, but rather as an added value.
+
+> The integration with the Language Server Protocol is created as an extension to the paid IntelliJ-based IDEs.
+> Therefore, plugins utilizing Language Server integration may not be available in Community releases of JetBrains products and Android Studio from Google.
+>
+{style="note"}
+
+Starting with the 2023.2 release cycle, the LSP API is publicly available as part of the IntelliJ Platform in the following IDEs:
+- IntelliJ IDEA Ultimate
+- WebStorm
+- PhpStorm
+- PyCharm Professional
+- DataSpell
+- RubyMine
+- CLion
+- Aqua
+- DataGrip
+- GoLand
+- Rider
+
+## Plugin Configuration
+
+To fully utilize the LSP API in a third-party plugin based on the [](tools_gradle_intellij_plugin.md), it is recommended to upgrade the Gradle IntelliJ Plugin to the latest version available.
+This plugin will attach the LSP API sources and code documentation to the project.
+
+As LSP became available in the 2023.2 EAP7 of IntelliJ-based IDEs, the plugin must target IntelliJ IDEA Ultimate `2023.2` or later.
+
+Example <path>build.gradle.kts</path> configuration:
+
+```kotlin
+plugins {
+  // ...
+  id("org.jetbrains.intellij") version "%gradle-intellij-plugin-version%"
+}
+
+intellij {
+  version = "%ijPlatform%"
+  type = "IU"
+}
+```
+
+For projects based on the [](plugin_github_template.md), update the Gradle IntelliJ Plugin to the latest version, and amend the <path>gradle.properties</path> file as follows:
+
+```
+platformType = IU
+platformVersion = %ijPlatform%
+```
+
+The <path>plugin.xml</path> configuration file needs to specify the dependency on the IntelliJ IDEA Ultimate module:
+
+```xml
+<idea-plugin>
+   <!-- ... -->
+   <depends>com.intellij.modules.ultimate</depends>
+</idea-plugin>
+```
+
+The LSP API sources are bundled in IntelliJ IDEA Ultimate and can be found within the <path>[IDE]/lib/src/src_lsp-openapi.zip</path> archive.
+
+## Supported Features
+
+The initial LSP support within the IntelliJ Platform covers the following features:
```

Comment:
I think now we may avoid the word `initial` and list the features as of 2023.3.

Since 2023.3:

- Intention actions ([textDocument/codeAction](https://microsoft.github.io/language-server-protocol/specification/#textDocument_codeAction))
- Code formatting ([textDocument/formatting](https://microsoft.github.io/language-server-protocol/specification/#textDocument_formatting))

Since 2023.3.1:

- Quick documentation ([textDocument/hover](https://microsoft.github.io/language-server-protocol/specification#textDocument_hover))

The first item could be about communication channel:

- Communication channels: STDIO (since 2023.2), sockets (since 2024.1)